### ISSUE TITLE
Hyperparameter advice

### ISSUE THREAD

**Issue by tbuckworth** [guojm14/ESPL#1]
I am attempting to use the EQL module for supervised learning (to learn the expression y = x^2) as a test, before implementing the module in other architectures for RL.

I keep ending up with a not very sparse symbolic formula, where the extracted symbolic expression is very large.

I was hoping for some advice on the hyperparameters for converging to a sufficiently sparse network, and thereby extracting a concise symbolic expression (hopefully x**2, or something close).

In particular how do the hyperparemeters `sample_num`, `hard_epoch`, `warmup_epoch`, `target_temp`, `target_ratio`, `spls` effect the resulting sparsity?

Is there something about my hyperparameter configuration that is obviously wrong and would not result in a sufficiently sparse network?

I have adjusted the code to take the `eql_args` as shown below and the `EQL` module has additional methods incorporating relevant code from the ESPL codebase.

```
    arch_index = 0
    init_op_list(arch_index)
    obs_dim = 1
    action_dim = 1
    cfg = dict(
        arch_index=arch_index,
        epochs=10000,
        data_size=1000,
        lr=1e-3,
        sample_num=2,
        hard_gum=True,
        data_scale=20,
    )
    eql_args = dict(
        target_ratio=0.001,
        spls=0.1,
        constrain_scale=1,
        l0_scale=0.01,
        bl0_scale=0,
        target_temp=0.2,
        warmup_epoch=0,
        hard_epoch=9000,
    )
    cfg.update(eql_args)
    epochs = cfg["epochs"]
    data_size = cfg["data_size"]
    lr = cfg["lr"]
    sample_num = cfg["sample_num"]
    hard_gum = cfg["hard_gum"]
    data_scale = cfg["data_scale"]

    num_inputs = obs_dim
    num_outputs = action_dim
    model = EQL(num_inputs, num_outputs, sample_num, hard_gum, **eql_args)
    
    optimizer = optim.Adam(model.parameters(), lr=lr)
    model.sample_sparse_constw(mode=0)
    obs = (np.random.random((data_size, 1)) - .5) * data_scale
    x = torch.FloatTensor(obs)
    y = x ** 2

    for epoch in range(epochs):
        y_hat = model.forward(x, mode=1)
        if sample_num > 1:
            y = y.unsqueeze(0).expand(sample_num, -1, -1).reshape(sample_num * data_size, -1)

        other_loss, sparse_loss, constrain_loss, regu_loss, l0_loss, bl0_loss = model.get_loss()
        total_loss = nn.MSELoss()(y, y_hat) + other_loss
        model.update_const()

        optimizer.zero_grad()
        total_loss.backward()
        nn.utils.clip_grad_norm_(model.parameters(), max_norm=0.5)
        optimizer.step()
        model.proj()
        model.set_temp_target_ratio(epoch)

    model.sample_sparse_constw(mode=0)
    y_hat_mode_0 = model.forward(x, mode=0)

    scores = model.scores.data
    constw_base = model.constw_base.data
    constb = model.constb.data
    constw = constw_base * ((scores > 0.5).float())
    sym_exp = str(printsymbolic(constw, constb, num_inputs, arch_index))
```

here is the added method to the `EQL` module:

```
    def set_temp_target_ratio(self, epoch):
        self.temp = 1 / ((1 - self.target_temp) * (
                1 - min(epoch, self.hard_epoch) / self.hard_epoch) + self.target_temp)
        clip_it = max(min(epoch, self.hard_epoch), self.warmup_epoch)
        self.target_ratio = self.target_ratio + (1 - self.target_ratio) * (
                1 - ((clip_it - self.warmup_epoch) / (self.hard_epoch - self.warmup_epoch)) ** 2)
```

**Comment by tbuckworth** [comment]
It occurs to me that my `MSELoss` may be much larger than the policy loss it is replacing in the original code. So perhaps I need to scale up the `other_loss`.

**Comment by jmguo96** [comment]
I haven't tried supervised learning, but I think it's similar for hyperparameters tuning.  In general, increasing spls, l0_scale, and bl0_scale will make the expression sparser. With spls having the biggest effect,  I did not adjust l0_scale and bl0_scale.  It is also possible to use a smaller target_ratio to reduce the minimum expression complexity.

**Comment by guojm14** [comment]
Another important thing is that I control target_ratio in the RL process (https://github.com/guojm14/ESPL/blob/c6ff1715a44d7d3428e72d7b4800d27c8a636083/ESPL/rlkit/core/symbolic_rl_algorithm.py#L95C1-L96C1) to keep the complexity down smoothly. This schedule is related to the setting of hard_epoch and the number of training epochs. If this is missing, the loss function that reduces the complexity may not work at all.

**Comment by jmguo96** [comment]
And If MSEloss is much larger than other losses, I recommend trying to increase the spls by an order of magnitude or two

**Comment by tbuckworth** [comment]
Thanks for your helpful advice.

I realise that I was incorrectly setting the target ratio.

I now use a new field `target_ratio_current`:
```
self.target_ratio_current = self.target_ratio + (1 - self.target_ratio) * (
                1 - ((clip_it - self.warmup_epoch) / (self.hard_epoch - self.warmup_epoch)) ** 2)
```

I have extracted a sparse expression modelling `x^2`.

This is the expression if of interest `-0.435*x0*(0.03 - 0.33*x0) + 2.04*(0.21*x0 - 0.e-2)*(0.27*x0 - 0.03) + 0.76*exp(0.43*x0) + 0.44*exp(0.48*x0) + 5.36 + 0.53*exp(-0.49*x0) + 0.67*exp(-0.4*x0)`.

I will adjust the hyperparameters according to your advice to see if I can find a closer approximation.

Thanks again!
